packages/@uppy/dashboard/src/components/FileCard/index.jsx

```diff
@@ -1,113 +1,104 @@
+    return typeof metaFields === 'function'
+      ? metaFields(files[fileCardFor])
+      : metaFields
+  }
 
-  constructor (props) {
-    super(props)
+  const file = files[fileCardFor]
+  const computedMetaFields = getMetaFields() ?? []
+  const showEditButton = canEditFile(file)
 
-    const file = this.props.files[this.props.fileCardFor]
-    const metaFields = this.getMetaFields() || []
+  const storedMetaData = {}
+  computedMetaFields.forEach((field) => {
+    storedMetaData[field.id] = file.meta[field.id] ?? ''
+  })
 
-    const storedMetaData = {}
-    metaFields.forEach((field) => {
-      storedMetaData[field.id] = file.meta[field.id] || ''
-    })
+  const [formState, setFormState] = useState(storedMetaData)
 
-    this.state = {
-      formState: storedMetaData,
-    }
-
-    this.form.id = nanoid()
-  }
-
-  // TODO(aduh95): move this to `UNSAFE_componentWillMount` when updating to Preact X+.
-  componentWillMount () { // eslint-disable-line react/no-deprecated
-    this.form.addEventListener('submit', this.handleSave)
-    document.body.appendChild(this.form)
-  }
-
-  componentWillUnmount () {
-    this.form.removeEventListener('submit', this.handleSave)
-    document.body.removeChild(this.form)
-  }
+  const handleSave = useCallback((ev) => {
+    ev.preventDefault()
+    saveFileCard(formState, fileCardFor)
+  }, [saveFileCard, formState, fileCardFor])
 
-  getMetaFields () {
-    return typeof this.props.metaFields === 'function'
-      ? this.props.metaFields(this.props.files[this.props.fileCardFor])
-      : this.props.metaFields
+  const updateMeta = (newVal, name) => {
+    setFormState({ [name]: newVal })
   }
 
-  updateMeta = (newVal, name) => {
-    this.setState(({ formState }) => ({
-      formState: {
-        ...formState,
-        [name]: newVal,
-      },
-    }))
+  const handleCancel = () => {
+    uppy.emit('file-editor:cancel', file)
+    toggleFileCard(false)
   }
 
-  handleSave = (e) => {
-    e.preventDefault()
-    const fileID = this.props.fileCardFor
-    this.props.saveFileCard(this.state.formState, fileID)
-  }
+  const [form] = useState(() => {
+    const formEl = document.createElement('form')
+    formEl.setAttribute('tabindex', '-1')
+    formEl.id = nanoid()
+    return formEl
+  })
 
-  handleCancel = () => {
-    const file = this.props.files[this.props.fileCardFor]
-    this.props.uppy.emit('file-editor:cancel', file)
-    this.props.toggleFileCard(false)
-  }
+  useEffect(() => {
+    document.body.appendChild(form)
+    return () => document.body.removeChild(form)
+  }, [form])
 
-  saveOnEnter = (ev) => {
-    if (ev.keyCode === 13) {
-      ev.stopPropagation()
-      ev.preventDefault()
-      const file = this.props.files[this.props.fileCardFor]
-      this.props.saveFileCard(this.state.formState, file.id)
-    }
-  }
+  useEffect(() => {
+    form.addEventListener('submit', handleSave)
+    return () => form.removeEventListener('submit', handleSave)
+  }, [form, handleSave])
```

Comment:
These two effects could be merged into one I think?

Reply:
@mifi in your prototype you separated them for some reason, I went with your way cause thought maybe it's better for performance, since the other one has extra dependency. Thoughts? :)